src/noise-models.lisp

```diff
@@ -0,0 +1,271 @@
+   (noise-position
+    :initarg :noise-position
+    :reader noise-position
+    :type noise-position
+    :documentation "Should application of a channel happen before or after the instruction?")
+   (name
+    :initarg :name
+    :reader name
+    :documentation "Name or descriptor for noise."))
+  (:default-initargs
+   :predicate-function nil
+   :priority 10
+   :noise-position ':after
+   :name +default-noise-predicate-name+)
+  (:documentation "A NOISE-PREDICATE describes a collection of program instructions. When used in a NOISE-RULE, the NOISE-PREDICATE describes which instructions should be matched by the noise described in the OPERATION-ELEMENTS. Optionally, the NOISE-PREDICATE can be given a NAME. If none is provided, use the default 'NOISE-NAME'."))
+
+
+(defmethod initialize-instance :after ((pred noise-predicate) &rest args)
+  ;; Assert that the NOISE-POSITION of the predicate is of type
+  ;; NOISE-POSITION, and that PRIORITY is of type INTEGER
+  (declare (ignore args))
+  (assert (typep (noise-position pred) 'noise-position))
+  (assert (typep (priority pred) 'integer))
+  (assert (typep (priority pred) 'noise-priority)))
+
+
+(defun make-noise-predicate (predicate-function priority noise-position &optional (name +DEFAULT-NOISE-PREDICATE-NAME+))
+  "Creates a NOISE-PREDICATE with the given PREDICATE-FUNCTION (function designator), PRIORITY, and NOISE-POSITION."
+  (make-instance 'noise-predicate :predicate-function predicate-function
+                                  :priority priority
+                                  :noise-position noise-position
+                                  :name name))
+
+
+(defun predicate-and (np1 np2)
+  "Logical AND of 2 noise predicates. The NOISE-POSITION is taken from NP1 and the priority is taken to be the min (smaller in value) priority between the two predicates."
+  (let* ((new-name (and-predicate-names (name np1) (name np2)))
+         (conjunction (alexandria:conjoin (predicate-function np1) (predicate-function np2)))
+         (priority (max (priority np1) (priority np2)))
+         (noise-position (noise-position np1)))
+    (make-noise-predicate conjunction priority noise-position new-name)))
+
+
+(defun predicate-or (np1 np2)
+  "Logical OR of 2 noise predicates. The NOISE-POSITION is taken from NP1 and the priority is taken to be the min (smaller in value) priority between the two predicates."
+  (let* ((new-name (or-predicate-names (name np1) (name np2)))
+         (disjunction (alexandria:disjoin (predicate-function np1) (predicate-function np2)))
+         (priority (max (priority np1) (priority np2)))
+         (noise-position (noise-position np1)))
+    (make-noise-predicate disjunction priority noise-position new-name)))
+
+
+(defmethod predicate-not (np)
+ "Logical NOT of a NOISE-PREDICATE NP. The PRIORITY of the predicate is 'inverted', by taking the difference between it and +maxpriority+."
```

Comment:
Microscopic nit: indent this documentation string by one additional space.

Reply:
Another nit: rename `np` to `noise-predicate` below and rewrite the docstring as:
```suggestion
 "Logical NOT of a NOISE-PREDICATE. The PRIORITY of the predicate is 'inverted', by taking the difference between it and +maxpriority+."
```
The rationale for doing this is that when you write code that calls `predicate-not`, Emacs will tell you the expected argument names and `NOISE-PREDICATE` says everything you need to know whereas `NP` is more cryptic. Same thing above with the other predicates.

This is just a suggestion and something for you to consider. It's not a deal-breaker for merging the code as far as I am concerned.